src/plugins/intel_cpu/src/nodes/executors/executor_implementation.hpp

```diff
@@ -0,0 +1,106 @@
+// Copyright (C) 2018-2022 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#pragma once
+
+#include <functional>
+#include <memory>
+
+#include "nodes/executors/executor.hpp"
+#include "nodes/executors/executor_config.hpp"
+
+namespace ov {
+namespace intel_cpu {
+
+// @todo Consider alternative of using template arguments instead of std::functions
+template <typename Attrs>
+class ExecutorImplementation {
+public:
+    ExecutorImplementation(
+        const char* name,
+        const ExecutorType type,
+        const OperationType operationType,
+        const ShapeTolerance shapeRelation,
+        std::function<bool(const executor::Config<Attrs>&)> isSupported,
+        std::function<std::pair<bool, executor::Config<Attrs>>(const executor::Config<Attrs>&)> isFullyCompliant,
+        std::function<bool(const executor::Config<Attrs>&)> isShapeSuitable,
+        std::function<ExecutorPtr(const executor::Config<Attrs>&,
+                                  const MemoryArgs& memory,
+                                  const ExecutorContext::CPtr context)> create)
+        : m_name(name),
+          m_type(type),
+          m_operationType(operationType),
+          m_shapeRelation(shapeRelation),
+          m_isSupported(std::move(isSupported)),
+          m_isFullyCompliant(std::move(isFullyCompliant)),
+          m_isShapeSuitable(std::move(isShapeSuitable)),
+          m_create(std::move(create)) {}
+
+    bool isSupported(const executor::Config<Attrs>& key) const {
+        if (m_isSupported) {
+            return m_isSupported(key);
+        }
+
+        return false;
+    }
+
+    std::pair<bool, executor::Config<Attrs>> isFullyCompliant(const executor::Config<Attrs>& key) const {
+        if (m_isFullyCompliant) {
+            return m_isFullyCompliant(key);
+        }
+
+        return {false, key};
+    }
+
+    bool isShapeSuitable(const executor::Config<Attrs>& key) const {
```

Comment:
I think we can extract the shapes and have only shapes as the arguments, but I think this would require having some intermediate storage which we probable want to avoid.

Reply:
We may request a vector of `std::ref` of pointers to the actual shapes so that to avoid temporal memory allocation. Having this signature, we extend the method scope beyond just shapes, which with no doubts will cause using this method beyond its purpose. If it's not what we want, better to allow only shapes as parameters.